webapp/client/socket.ts

```diff
@@ -0,0 +1,73 @@
+import types from '@/bindings/bindings'
+
+export type ServerCallback = (socket:Socket) => void
+export type OnMessageCallback = (message:types.ServerMessage, socket:Socket) => void
+
+export interface UpdateObject {
+    callback: ServerCallback | undefined
+    timeInterval: number
+}
+
+export default class Socket {
+    connection : WebSocket;
+    onOpen : ServerCallback
+    updateInterval : number
+    reconnectionTimout = 5000
+    updateObject : UpdateObject
+    onMessage : OnMessageCallback
+
+    // has to be run only in useEffect because we need the window to exist
+    // to create a websocket and run updates
+    constructor(serverUrl : string, onOpen : ServerCallback, updateObject : UpdateObject, onMessage:OnMessageCallback) {
+        this.connection = new WebSocket(serverUrl);
+        this.onMessage = onMessage
+        this.updateObject = updateObject
+        this.onOpen = onOpen
+
+        this.setSocketHandlers()
+    }
+
+    setSocketHandlers() {
+        this.connection.onopen = () => {
+            this.onOpen(this)
+            console.info("Successfuly connected.")
+            this.update(this.updateObject)
+        }
+        this.connection.onmessage = (event : MessageEvent) => {
+            // bad implementation by ts, will read that: 
+            // https://dev.to/codeprototype/safely-parsing-json-to-a-typescript-interface-3lkj 
+            // some day
+            this.onMessage(JSON.parse(event.data) as types.ServerMessage, this)
+        }
+        this.connection.onclose = () => this.reconnect()
+    }
+
+    send(message : types.ClientMessage) {
+        // might accidentally send a message while connecting
+        if (this.connection.readyState == WebSocket.OPEN) {
+            this.connection.send(JSON.stringify(message))
+        }
+    }
+
+    update(updateObject : UpdateObject) {
+        this.updateInterval = window.setInterval(() => {
+            updateObject.callback?.(this)
+        }, 
+        updateObject.timeInterval)
+    }
+
+    stopUpdating() {
+        clearInterval(this.updateInterval)
+    }
+
+    reconnect() {
+        console.info('Attempt to reconnect has begun.')
```

Comment:
Im pretty sure you also need to update the `GameState` when you reconnect, maybe add an `OnConnectionClosedCallback`?

Reply:
`onOpen` is called after the socket closes because of the reconnection method. `onOpen` sets the `GameState` to `inQuery`, so it's updated automatically.

But adding a `waiting` state could be a great addition.
Added an issue for that: #61